File: watch/Shared/WatchProtocol.swift
import Foundation

enum WatchLocale {
  static let supported = [
    "en", "zh-Hans", "zh-Hant", "fr", "de", "hi",
    "it", "ja", "ko", "es", "th", "vi",
  ]
  static let defaultFormattingIdentifiers = [
    "en": "en-AU", "zh-Hans": "zh-CN", "zh-Hant": "zh-TW",
    "fr": "fr-FR", "de": "de-DE", "hi": "hi-IN", "it": "it-IT",
    "ja": "ja-JP", "ko": "ko-KR", "es": "es-ES", "th": "th-TH",
    "vi": "vi-VN",
  ]

  static func canonical(_ value: String?) -> String? {
    guard let value else { return nil }
    let normalized = value.replacingOccurrences(of: "_", with: "-").lowercased()
    if normalized.hasPrefix("zh-") || normalized == "zh" {
      return normalized.contains("hant") || normalized.contains("-tw") ||
        normalized.contains("-hk") || normalized.contains("-mo") ? "zh-Hant" : "zh-Hans"
    }
    let language = normalized.split(separator: "-").first.map(String.init) ?? normalized
    return supported.first { $0.lowercased() == language }
  }

  static func resolve(locale: String?, language: String?, system: Locale = .current) -> String {
    if let selected = canonical(locale) { return selected }
    if language == "zh" { return "zh-Hans" }
    if language == "en" { return "en" }
    return canonical(system.identifier) ?? "en"
  }

  static func canonicalFormatting(_ value: String?, for catalog: String) -> String? {
    guard let value else { return nil }
    let normalized = value.trimmingCharacters(in: .whitespacesAndNewlines)
      .replacingOccurrences(of: "_", with: "-")
    guard !normalized.isEmpty, normalized.count <= 64,
          canonical(normalized) == catalog else { return nil }
    return normalized
  }

  static func resolveFormatting(locale: String?, language: String?, formattingLocale: String?,
    system: Locale = .current) -> String {
    let catalog = resolve(locale: locale, language: language, system: system)
    if let selected = canonicalFormatting(formattingLocale, for: catalog) { return selected }
    let hasPhoneSelection = canonical(locale) != nil || language == "zh" || language == "en"
    if !hasPhoneSelection,
       canonical(system.identifier) == catalog,
       let selected = canonicalFormatting(system.identifier, for: catalog) { return selected }
    return defaultFormattingIdentifiers[catalog] ?? "en-AU"
  }
}

enum WatchClock {
  static func string(_ date: Date = Date()) -> String {
    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    return formatter.string(from: date)
  }
  static func date(_ value: String) -> Date? {
    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    return formatter.date(from: value) ?? ISO8601DateFormatter().date(from: value)
  }
}

struct WatchEntry: Codable, Equatable, Identifiable {
  let id: String
  let type: String
  let start: String
  var end: String?
  var feedRunning: Bool?
  var amount: Double?
  var feedKind: String?
  var diaperKind: String?
  var version: String?
  var canControl: Bool?
  var pendingOperationId: String?
  var note: String = ""

  var isActive: Bool { (type == "sleep" && end == nil) || (type == "feed" && feedRunning == true) }
  var startedAt: Date { WatchClock.date(start) ?? .distantPast }
  var isBottle: Bool { feedKind == "formula" || feedKind == "expressed" }
  var commandEntry: WatchEntry {
    var value = self
    value.version = nil
    value.canControl = nil
    value.pendingOperationId = nil
    return value
  }
}

struct WatchContext: Codable {
  struct Profile: Codable { let name: String; let birthDate: String }
  struct Totals: Codable { let feedMl: Double; let feedCount: Int; let diaperCount: Int; let sleepMinutes: Double }
  let schemaVersion: Int
  let workspaceKey: String
  let bridgeId: String
  let generation: Int
  let sequence: Int
  let status: String
  let mode: String?
  let expiresAt: String?
  let publishedAt: String?
  let profile: Profile?
  let entries: [WatchEntry]?
  let language: String?
  let totals: Totals?
  var totalsDate: String?
  // Completed sleep intervals, merged and bounded by the phone, in Unix seconds.
  // Older phones omit them; in that case sleep totals wait for confirmation.
  var sleepRanges: [[Double]]?
  // Canonical app-selected locale. Older phones only send language (en|zh),
  // so this remains optional without changing the version 1 wire contract.
  var locale: String? = nil
  // Regional formatting locale. Optional so previously persisted version 1
  // contexts continue to decode and use the catalog's stable default region.
  var formattingLocale: String? = nil
  // Keep an authenticated read/conflict context usable if the phone pauses new
  // Watch recording. Older phone contexts omit this and remain enabled.
  var recordingEnabled: Bool? = nil

  func isValid(at now: Date) -> Bool {
    guard schemaVersion == 1, status == "ready", !workspaceKey.isEmpty,
          let expiresAt, let expiry = WatchClock.date(expiresAt),
          let publishedAt, let published = WatchClock.date(publishedAt) else { return false }
    return now >= published.addingTimeInterval(-60) && now < min(expiry, published.addingTimeInterval(24 * 60 * 60))
  }
  func supersedes(_ old: WatchContext?) -> Bool {
    guard let old else { return true }
    if bridgeId != old.bridgeId { return true }
    return generation > old.generation || (generation == old.generation && sequence > old.sequence && workspaceKey == old.workspaceKey)
  }
}

struct WatchCommand: Codable, Identifiable {
  let schemaVersion: Int
  let commandId: String
  let recordId: String
  let workspaceKey: String
  let bridgeId: String
  let generation: Int
  let snapshotSequence: Int
  let createdAt: String
  let kind: String
  var entry: WatchEntry?
  var stoppedAt: String?
  var amount: Double?
  var baseVersion: String?
  var expectedEntry: WatchEntry?
  var dependsOn: String?
  // A conflict resolution is a new immutable command that refers to the
  // rejected command the person reviewed; it never mutates that old command.
  var resolution: String?
  var conflictOperationId: String?
  var id: String { commandId }

  func json() throws -> String {
    let encoder = JSONEncoder()
    encoder.outputFormatting = [.sortedKeys]
    return String(decoding: try encoder.encode(self), as: UTF8.self)
  }
}

struct WatchConflict: Codable {
  let currentVersion: String
  let currentEditedBy: String
  let currentEntry: WatchEntry
  let proposedEntry: WatchEntry
  let canReplace: Bool
}

struct WatchReceipt: Codable {
  let schemaVersion: Int
  let commandId: String
  let recordId: String?
  let workspaceKey: String
  var bridgeId: String?
  let generation: Int
  let status: String
  let error: String?
  let contextSequence: Int?
  var conflict: WatchConflict?
}

struct WatchOutboxItem: Codable, Identifiable {
  let command: WatchCommand
  var receivedByPhone = false
  var receipt: WatchReceipt?
  var projectionReconciled = false
  var id: String { command.commandId }
  var terminal: Bool { ["saved", "shared", "rejected"].contains(receipt?.status ?? "") }
}

// Watch-local input seed, never part of a record or command sent to the phone.
struct WatchMilkAmountDraft: Codable {
  let recordId: String
  let workspaceKey: String
  let bridgeId: String
  let generation: Int
  let amount: Int

  func matches(_ entry: WatchEntry, context: WatchContext?) -> Bool {
    recordId == entry.id && workspaceKey == context?.workspaceKey &&
      bridgeId == context?.bridgeId && generation == context?.generation
  }
}

struct WatchDisk: Codable {
  var context: WatchContext?
  var outbox: [WatchOutboxItem] = []
  var retiredBridgeIds: [String] = []
  var milkAmountDraft: WatchMilkAmountDraft?
  var summaryBase: WatchContext?
  var summaryCommandIds: [String]?

  mutating func append(_ command: WatchCommand) {
    // Pin the last known phone totals BEFORE adding the first local operation.
    // Keep this baseline until the entire batch is reflected by phone receipts
    // AND a corresponding snapshot, regardless of transport arrival order.
    if summaryBase == nil {
      summaryBase = context
      summaryCommandIds = []
    }
    summaryCommandIds?.append(command.id)
    outbox.append(WatchOutboxItem(command: command))
  }

  func summary(at now: Date, calendar: Calendar = .current) -> WatchSummary? {
    guard let context, context.isValid(at: now) else { return nil }
    let base = summaryBase ?? context
    let start = calendar.startOfDay(for: now)
    guard let end = calendar.date(byAdding: .day, value: 1, to: start) else { return nil }
    let formatter = DateFormatter()
    formatter.calendar = Calendar(identifier: .gregorian)
    formatter.timeZone = calendar.timeZone
    formatter.locale = Locale(identifier: "en_US_POSIX")
    formatter.dateFormat = "yyyy-MM-dd"
    let isToday = base.totalsDate == formatter.string(from: now)
    var milk = isToday ? base.totals?.feedMl ?? 0 : 0
    var feeds = isToday ? base.totals?.feedCount ?? 0 : 0
    var nappies = isToday ? base.totals?.diaperCount ?? 0 : 0
    let ids = Set(summaryCommandIds ?? [])
    let batch = outbox.filter { ids.contains($0.id) && $0.command.workspaceKey == context.workspaceKey &&
      $0.command.bridgeId == context.bridgeId && $0.command.generation == context.generation }
    let rejected = Set(outbox.filter { $0.receipt?.status == "rejected" }.map(\.id))
    var records: [String: WatchEntry] = [:]
    var sleepRanges = isToday ? base.sleepRanges ?? [] : []
    var hasLocal = false
    for item in batch {
      let command = item.command
      guard !rejected.contains(item.id), !rejected.contains(command.dependsOn ?? "") else { continue }
      if command.kind == "create", let entry = command.entry {
        records[entry.id] = entry
        hasLocal = true
      } else if let expected = command.expectedEntry,
                let stopped = command.stoppedAt.flatMap(WatchClock.date) {
        var entry = records[command.recordId] ?? expected
        if command.kind == "finish-feed" {
          // A phone-started feed is already counted; only its amount changes.
          if isToday && records[entry.id] == nil && base.entries?.contains(where: { $0.id == entry.id }) == true &&
              entry.startedAt >= start && entry.startedAt < end {
            feeds -= 1
            milk -= entry.amount ?? 0
          }
          entry.feedRunning = false
          entry.amount = entry.isBottle ? command.amount ?? 0 : nil
          entry.end = command.stoppedAt
          records[entry.id] = entry
        } else if command.kind == "finish-sleep" {
          records.removeValue(forKey: entry.id)
          if stopped.timeIntervalSince(entry.startedAt) >= 60 {
            entry.end = command.stoppedAt
            records[entry.id] = entry
          }
        }
        hasLocal = true
      }
    }
    for entry in records.values {
      if entry.startedAt >= start && entry.startedAt < end {
        if entry.type == "feed" { feeds += 1; milk += entry.amount ?? 0 }
        if entry.type == "diaper" { nappies += 1 }
      }
      if entry.type == "sleep", let finished = entry.end.flatMap(WatchClock.date) {
        sleepRanges.append([max(start, entry.startedAt).timeIntervalSince1970,
          min(end, finished).timeIntervalSince1970])
      }
    }
    // Merge with the phone's completed intervals, not just add minutes: a
    // backfilled sleep can overlap a Watch timer. Missing/oversized old metadata
    // must not manufacture precision; retain the phone sleep total until synced.
    let sleep = !isToday || base.sleepRanges != nil
      ? WatchSummary.sleepMinutes(sleepRanges)
      : base.totals?.sleepMinutes ?? 0
    return WatchSummary(totals: .init(feedMl: max(0, milk), feedCount: max(0, feeds),
      diaperCount: max(0, nappies), sleepMinutes: sleep), phoneUpdatedAt: base.publishedAt,
      needsPhoneUpdate: !isToday, includesLocalChanges: hasLocal)
  }

  func initialMilkAmount(for entry: WatchEntry) -> Int {
    if let draft = milkAmountDraft, draft.matches(entry, context: context), (0...2_000).contains(draft.amount) {
      return draft.amount
    }
    // Older/phone-started timers have no Watch draft. Zero is their unfinished
    // record placeholder, not a previously selected consumed amount.
    if let amount = entry.amount, amount.isFinite, amount > 0, amount <= 2_000 {
      return Int(amount.rounded())
    }
    return 120
  }

  mutating func rememberMilkAmount(_ amount: Int, for entry: WatchEntry) {
    guard entry.isBottle, entry.feedRunning == true, (0...2_000).contains(amount), let context else { return }
    milkAmountDraft = WatchMilkAmountDraft(recordId: entry.id, workspaceKey: context.workspaceKey,
      bridgeId: context.bridgeId, generation: context.generation, amount: amount)
  }

  func visibleEntries() -> [WatchEntry] {
    guard let context else { return [] }
    var values = context.entries ?? []
    for item in outbox where item.command.bridgeId == context.bridgeId && item.command.workspaceKey == context.workspaceKey && item.command.generation == context.generation {
      guard item.receipt?.status != "rejected", !item.projectionReconciled else { continue }
      let command = item.command
      if command.kind == "create", let entry = command.entry {
        values.removeAll { $0.id == entry.id }
        if entry.isActive { values.append(entry) }
      } else if command.kind == "finish-sleep" || command.kind == "finish-feed" {
        values.removeAll { $0.id == command.recordId }
      }
    }
    return values.filter(\.isActive)
  }

  mutating func acceptContext(_ value: WatchContext) {
    guard value.schemaVersion == 1, !retiredBridgeIds.contains(value.bridgeId), value.supersedes(context) else { return }
    if let old = context, old.bridgeId != value.bridgeId { retiredBridgeIds.append(old.bridgeId) }
    if let base = summaryBase, value.bridgeId != base.bridgeId ||
        value.workspaceKey != base.workspaceKey || value.generation != base.generation {
      summaryBase = nil
      summaryCommandIds = nil
    }
    context = value
    reconcileProjection()
  }

  mutating func reconcileProjection() {
    if let draft = milkAmountDraft,
       !visibleEntries().contains(where: { $0.isBottle && draft.matches($0, context: context) }) {
      milkAmountDraft = nil
    }
    guard let value = context, value.status == "ready" else { return }
    for index in outbox.indices {
      let item = outbox[index]
      guard !item.projectionReconciled,
            item.command.workspaceKey == value.workspaceKey,
            item.command.bridgeId == value.bridgeId,
            item.command.generation == value.generation,
            let receipt = item.receipt, receipt.status != "rejected",
            value.sequence >= (receipt.contextSequence ?? Int.max) else { continue }
      let present = value.entries?.contains { $0.id == item.command.recordId && $0.isActive } ?? false
      if item.command.kind == "resolve-conflict" {
        outbox[index].projectionReconciled = item.terminal
      } else if item.command.kind == "create" {
        // Only replace an active local start once the phone snapshot contains it.
        outbox[index].projectionReconciled = item.command.entry?.isActive != true || present ||
          (item.terminal && value.sequence > (receipt.contextSequence ?? Int.max))
      } else {
        outbox[index].projectionReconciled = !present
      }
    }
    // A terminal resolution supersedes the rejected command it reviewed. If
    // replacement raced with another edit, the resolution command itself is
    // rejected with a fresh conflict and remains available for another review.
    let resolved = Set(outbox.compactMap { item in
      item.command.kind == "resolve-conflict" && item.terminal
        ? item.command.conflictOperationId : nil
    })
    outbox.removeAll { resolved.contains($0.id) }
    // Retain bounded terminal receipt history, not unsynced work. Quarantined old
    // generations are never replayed or rendered under the new workspace.
    if let ids = summaryCommandIds, ids.allSatisfy({ id in
      outbox.contains { $0.id == id && ($0.projectionReconciled || $0.receipt?.status == "rejected") }
    }) {
      summaryBase = nil
      summaryCommandIds = nil
    }
    let pinned = Set(summaryCommandIds ?? [])
    let finished = outbox.filter { !pinned.contains($0.id) && ["saved", "shared"].contains($0.receipt?.status ?? "") && $0.projectionReconciled }
    let removable = Set(finished.dropLast(100).map(\.id))
    outbox.removeAll { removable.contains($0.id) }
  }
}

struct WatchSummary {
  let totals: WatchContext.Totals
  let phoneUpdatedAt: String?
  let needsPhoneUpdate: Bool
  let includesLocalChanges: Bool

  static func sleepMinutes(_ ranges: [[Double]]) -> Double {
    let sorted = ranges.filter { $0.count == 2 && $0[0].isFinite && $0[1].isFinite && $0[1] > $0[0] }
      .sorted { $0[0] < $1[0] }
    var stop = -Double.infinity
    var seconds = 0.0
    for range in sorted {
      seconds += max(0, range[1] - max(stop, range[0]))
      stop = max(stop, range[1])
    }
    return seconds / 60
  }
}

File: src/watchProtocol.ts
import { type Entry, type State, validateEntry, validateState } from "./domain";
import { finishLiveSleep } from "./sleepTimer";
import { finishFeed } from "./feedFinish";
import type { SupportedLocale } from "./locales";

export type WatchCommand = {
  schemaVersion: 1;
  commandId: string;
  recordId: string;
  workspaceKey: string;
  generation: number;
  bridgeId?: string;
  snapshotSequence?: number;
  createdAt: string;
  kind: "create" | "finish-sleep" | "finish-feed" | "resolve-conflict";
  entry?: Entry;
  stoppedAt?: string;
  amount?: number;
  baseVersion?: string;
  expectedEntry?: Entry;
  dependsOn?: string;
  resolution?: "discard" | "replace";
  conflictOperationId?: string;
};
export type WatchConflict = {
  currentVersion: string;
  currentEditedBy: string;
  currentEntry: Entry;
  proposedEntry: Entry;
  canReplace: boolean;
};
export type WatchReceipt = {
  schemaVersion: 1;
  commandId: string;
  recordId: string;
  workspaceKey: string;
  generation: number;
  bridgeId?: string;
  status: "saved" | "pending" | "shared" | "rejected";
  error?: string;
  conflict?: WatchConflict;
};
export type WatchLedger = Record<
  string,
  { fingerprint: string; receipt: WatchReceipt }
>;
export type WatchContext = {
  schemaVersion: 1;
  workspaceKey: string;
  mode: "personal" | "family";
  status: "ready" | "unavailable";
  expiresAt: string;
  /** Legacy compatibility field consumed by older Watch/widget builds. */
  language: "zh" | "en";
  /** Canonical locale consumed by current Watch/widget builds. */
  locale?: SupportedLocale;
  /** Regional formatting locale; optional for version 1 compatibility. */
  formattingLocale?: string;
  /** New recording can be paused without hiding already-issued conflicts. */
  recordingEnabled?: boolean;
  profile: { name: string; birthDate: string };
  entries: (Entry & {
    version?: string;
    canControl?: boolean;
    pendingOperationId?: string;
  })[];
  totals: {
    feedMl: number;
    feedCount: number;
    diaperCount: number;
    sleepMinutes: number;
  };
  totalsDate: string;
  sleepRanges?: [number, number][];
  // Read-only widget access is independent from the Watch recording rollout.
  widgetTotals?: { feedMl: number; diaperCount: number; sleepMinutes: number };
  bridgeId?: string;
  generation?: number;
  sequence?: number;
  invalidate?: boolean;
};
const uuid = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
const instant = (value: unknown) =>
  typeof value === "string" &&
  /^\d{4}-\d{2}-\d{2}T/.test(value) &&
  Number.isFinite(Date.parse(value));
export function parseWatchCommand(raw: string): WatchCommand {
  if (raw.length > 16384) throw new Error("invalid_watch_command");
  const c = JSON.parse(raw) as WatchCommand;
  if (
    !c ||
    c.schemaVersion !== 1 ||
    !uuid.test(c.commandId) ||
    typeof c.recordId !== "string" ||
    !c.recordId.trim() ||
    c.recordId.length > 128 ||
    typeof c.workspaceKey !== "string" ||
    !c.workspaceKey ||
    c.workspaceKey.length > 2048 ||
    !Number.isSafeInteger(c.generation) ||
    c.generation < 1 ||
    (c.bridgeId !== undefined && !uuid.test(c.bridgeId)) ||
    !instant(c.createdAt) ||
    !["create", "finish-sleep", "finish-feed", "resolve-conflict"].includes(
      c.kind,
    ) ||
    (c.dependsOn !== undefined && !uuid.test(c.dependsOn)) ||
    (c.baseVersion !== undefined &&
      (typeof c.baseVersion !== "string" || c.baseVersion.length > 128))
  )
    throw new Error("invalid_watch_command");
  if (c.kind === "resolve-conflict") {
    if (
      !["discard", "replace"].includes(c.resolution ?? "") ||
      !c.conflictOperationId ||
      !uuid.test(c.conflictOperationId) ||
      c.conflictOperationId === c.commandId ||
      c.entry !== undefined ||
      c.stoppedAt !== undefined ||
      c.amount !== undefined ||
      c.expectedEntry !== undefined ||
      c.dependsOn !== undefined ||
      (c.resolution === "replace"
        ? !c.baseVersion
        : c.baseVersion !== undefined)
    )
      throw new Error("invalid_watch_command");
  } else if (c.resolution !== undefined || c.conflictOperationId !== undefined)
    throw new Error("invalid_watch_command");
  else if (c.kind === "create") {
    c.entry = validateEntry(c.entry);
    if (
      c.entry.id !== c.recordId ||
      !["feed", "diaper", "sleep"].includes(c.entry.type) ||
      c.entry.note !== ""
    )
      throw new Error("invalid_watch_command");
  } else if (!instant(c.stoppedAt)) throw new Error("invalid_watch_command");
  if (c.expectedEntry !== undefined)
    c.expectedEntry = validateEntry(c.expectedEntry);
  return c;
}

function validateWatchConflict(value: unknown): WatchConflict {
  const conflict = value as WatchConflict;
  if (
    !conflict ||
    typeof conflict !== "object" ||
    typeof conflict.currentVersion !== "string" ||
    !conflict.currentVersion ||
    conflict.currentVersion.length > 128 ||
    typeof conflict.currentEditedBy !== "string" ||
    !conflict.currentEditedBy ||
    conflict.currentEditedBy.length > 80 ||
    typeof conflict.canReplace !== "boolean"
  )
    throw new Error("local_data_invalid");
  const currentEntry = validateEntry(conflict.currentEntry);
  const proposedEntry = validateEntry(conflict.proposedEntry);
  if (
    (currentEntry.type !== "feed" && currentEntry.type !== "sleep") ||
    currentEntry.type !== proposedEntry.type ||
    currentEntry.id !== proposedEntry.id
  )
    throw new Error("local_data_invalid");
  return { ...conflict, currentEntry, proposedEntry };
}
function canonical(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(canonical);
  if (value && typeof value === "object")
    return Object.fromEntries(
      Object.entries(value)
        .sort(([a], [b]) => a.localeCompare(b))
        .map(([k, v]) => [k, canonical(v)]),
    );
  return value;
}
export const watchFingerprint = (command: WatchCommand) =>
  JSON.stringify(canonical(command));
export function watchReceipt(
  command: WatchCommand,
  status: WatchReceipt["status"],
  error?: string,
): WatchReceipt {
  return {
    schemaVersion: 1,
    commandId: command.commandId,
    recordId: command.recordId,
    workspaceKey: command.workspaceKey,
    generation: command.generation,
    ...(command.bridgeId ? { bridgeId: command.bridgeId } : {}),
    status,
    ...(error ? { error } : {}),
  };
}
export function previousWatchReceipt(
  ledger: WatchLedger | undefined,
  command: WatchCommand,
): WatchReceipt | undefined {
  const previous = ledger?.[command.commandId];
  if (previous && previous.fingerprint !== watchFingerprint(command))
    throw new Error("operation_reused");
  return previous?.receipt;
}
export function addWatchReceipt(
  ledger: WatchLedger | undefined,
  command: WatchCommand,
  receipt: WatchReceipt,
): WatchLedger {
  previousWatchReceipt(ledger, command);
  if (!ledger?.[command.commandId] && Object.keys(ledger ?? {}).length >= 10000)
    throw new Error("watch_ledger_full");
  return {
    ...ledger,
    [command.commandId]: { fingerprint: watchFingerprint(command), receipt },
  };
}
export function validateWatchLedger(value: unknown): WatchLedger | undefined {
  if (value === undefined) return undefined;
  if (
    !value ||
    typeof value !== "object" ||
    Array.isArray(value) ||
    Object.keys(value).length > 10000
  )
    throw new Error("local_data_invalid");
  for (const [id, item] of Object.entries(value as WatchLedger)) {
    if (!item || typeof item.fingerprint !== "string")
      throw new Error("local_data_invalid");
    const command = parseWatchCommand(item.fingerprint);
    const receipt = item.receipt;
    if (
      command.commandId !== id ||
      !receipt ||
      receipt.commandId !== id ||
      receipt.schemaVersion !== 1 ||
      receipt.recordId !== command.recordId ||
      receipt.workspaceKey !== command.workspaceKey ||
      receipt.generation !== command.generation ||
      receipt.bridgeId !== command.bridgeId ||
      !["saved", "pending", "shared", "rejected"].includes(receipt.status) ||
      (receipt.conflict !== undefined && receipt.status !== "rejected")
    )
      throw new Error("local_data_invalid");
    if (receipt.conflict !== undefined)
      receipt.conflict = validateWatchConflict(receipt.conflict);
  }
  return value as WatchLedger;
}
export function checkWatchTime(
  command: WatchCommand,
  state: State,
  now = Date.now(),
) {
  const timestamps = [
    command.createdAt,
    command.entry?.start,
    command.entry?.end,
    command.stoppedAt,
  ].filter((v): v is string => !!v);
  if (
    timestamps.some(
      (v) => !Number.isFinite(Date.parse(v)) || Date.parse(v) > now + 60000,
    )
  )
    throw new Error("invalid_record_time");
  const start = command.entry?.start;
  if (start && state.profile.birthDate) {
    const date = new Date(start);
    const localDate = `${date.getFullYear()}-${String(date.getMonth() + 1).padStart(2, "0")}-${String(date.getDate()).padStart(2, "0")}`;
    if (localDate < state.profile.birthDate)
      throw new Error("before_birth_date");
  }
}
export function watchEntryChange(
  state: State,
  command: WatchCommand,
  now = Date.now(),
): Entry | null {
  checkWatchTime(command, state, now);
  if (command.kind === "resolve-conflict")
    throw new Error("invalid_watch_command");
  const entry = state.entries.find((e) => e.id === command.recordId);
  if (command.kind === "create") {
    if (entry) throw new Error("record_changed");
    const created = validateEntry(command.entry);
    if (
      created.type === "sleep" &&
      !created.end &&
      state.entries.some((e) => e.type === "sleep" && !e.end)
    )
      throw new Error("running_sleep");
    if (created.feedRunning && state.entries.some((e) => e.feedRunning))
      throw new Error("running_feed");
    return created;
  }
  if (!entry)
    throw new Error(
      command.dependsOn ? "watch_dependency_pending" : "record_changed",
    );
  if (
    entry.end ||
    (command.kind === "finish-sleep"
      ? entry.type !== "sleep"
      : entry.type !== "feed" || !entry.feedRunning)
  )
    throw new Error("record_changed");
  if (
    command.expectedEntry &&
    (command.expectedEntry.id !== entry.id ||
      command.expectedEntry.start !== entry.start ||
      command.expectedEntry.type !== entry.type ||
      command.expectedEntry.feedKind !== entry.feedKind)
  )
    throw new Error("record_changed");
  return command.kind === "finish-sleep"
    ? finishLiveSleep(entry, command.stoppedAt!)
    : finishFeed(entry, command.stoppedAt!, command.amount);
}
export function applyPersonalWatchCommand(
  state: State,
  ledger: WatchLedger | undefined,
  command: WatchCommand,
  now = Date.now(),
) {
  const prior = previousWatchReceipt(ledger, command);
  if (prior) return { state, ledger: ledger!, receipt: prior };
  const entry = watchEntryChange(state, command, now);
  const next = validateState({
    ...state,
    entries: [
      ...state.entries.filter((e) => e.id !== command.recordId),
      ...(entry ? [entry] : []),
    ],
  });
  const receipt = watchReceipt(command, "saved");
  return {
    state: next,
    ledger: addWatchReceipt(ledger, command, receipt),
    receipt,
  };
}
